tonic-examples/routeguide-tutorial.md

```diff
@@ -0,0 +1,833 @@
+
+```shell
+$ mkdir data && touch data/route_guide_db.json
+$ touch src/data.rs
+```
+
+You can find our example json data in [tonic-examples/data/route_guide_db.json][route-guide-db] and
+the corresponding `data` module to load and deserialize it in
+[tonic-examples/routeguide/data.rs][data-module].
+
+**Note:** If you are following along, you'll need to change the data file's path  from
+`tonic-examples/data/route_guide_db.json` to `data/route_guide_db.json`.
+
+Next, we need to implement `Hash` and `Eq` for `Point`, so we can use point values as map keys:
+
+```rust
+use std::hash::{Hasher, Hash};
+```
+
+```rust
+impl Hash for Point {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        self.latitude.hash(state);
+        self.longitude.hash(state);
+    }
+}
+
+impl Eq for Point {}
+
+```
+
+Lastly, we need implement two helper functions: `in_range` and `calc_distance`. We'll use them
+when performing feature lookups. You can find them in
+[tonic-examples/src/routeguide/server.rs][in-range-fn].
+
+[route-guide-db]: https://github.com/hyperium/tonic/blob/master/tonic-examples/data/route_guide_db.json
+[data-module]: https://github.com/hyperium/tonic/blob/master/tonic-examples/src/routeguide/data.rs
+[in-range-fn]: https://github.com/hyperium/tonic/blob/master/tonic-examples/src/routeguide/server.rs#L188
+
+#### Request and Response types
+All our service methods receive a `tonic::Request<T>` and return a
+`Result<tonic::Response<T>, tonic::Status>`. The concrete type of `T` depends on how our methods
+are declared in our *service* `.proto` definition. It can be either:
+
+- A single value, e.g `Point`, `Rectangle`, or even a message type that includes a repeated field.
+- A stream of values, e.g. `impl Stream<Item = Result<Feature, tonic::Status>>`.
+
+#### Simple RPC
+Let's look at the simplest method first, `get_feature`, which just gets a `tonic::Request<Point>` 
+from the client and tries to find a feature at the given `Point`. If no feature is found, it returns
+an empty one.
+
+```rust
+async fn get_feature(&self, request: Request<Point>) -> Result<Response<Feature>, Status> {
+    for feature in &self.features[..] {
+        if feature.location.as_ref() == Some(request.get_ref()) {
+            return Ok(Response::new(feature.clone()));
+        }
+    }
+
+    let response = Response::new(Feature {
+        name: "".to_string(),
+        location: None,
+    });
+
+    Ok(response)
+}
+```
+
+
+#### Server-side streaming RPC
+Now let's look at one of our streaming RPCs. `list_features` is a server-side streaming RPC, so we
+need to send back multiple `Feature`s to our client.
+
+```rust
+type ListFeaturesStream = mpsc::Receiver<Result<Feature, Status>>;
+
+async fn list_features(
+    &self,
+    request: Request<Rectangle>,
+) -> Result<Response<Self::ListFeaturesStream>, Status> {
+    let (mut tx, rx) = mpsc::channel(4);
+    let features = self.features.clone();
+
+    tokio::spawn(async move {
```

Comment:
This would be a great way to use the `async-stream` crate right? Maybe mention it here?

Reply:
async-stream is used on the bidirectional streaming example and we talk abut it briefly. Do you think we should use it here too or explain in a bit more detail what it does?